File: R/mutation_rates.R
# Calculate mutation rates
#
# Loads gene-based mutation rates, in order to determine the expected number of
# mutations per gene, given the number of studied probands and adjusts for
# sex-chromosome transmissions.

#' gets mutation rates from the gene-based datasets
#'
#' This defaults to the gene-based mutation rates from Nature Genetics
#' 46:944-950 (2014) doi:10.1038/ng.3050, but we can pass in other gene-based
#' mutation rate datasets.
#'
#' @param trios list of male and female proband counts in the dataset
#' @param rates optionally specify a dataframe containing per-gene mutation
#'    rates, or NULL
#' @export
#'
#' @return a dataframe of mutation rates for genes under different mutation
#'     classes.
get_gene_based_mutation_rates <- function(trios, rates=NULL) {
    
    # Get the number of autosomal transmissions, so we can estimate the expected
    # number of mutations given the number of potential transmissions.
    autosomal = 2 * (trios$male + trios$female)
    
    # if we haven't passed in a rates dataset, default to the gene rates from
    # Samocha et al., Nature Genetics 46:944-950.
    if (is.null(rates)) { rates = mupit::gene_rates }
    
    # add chromosome annotation to the rates, if it isn't already included,
    # so that we can later correct for chrX transmission differences
    if (!("chrom" %in% names(rates))) {
        rates = merge(rates, mupit::gene_info, by="hgnc", all.x=TRUE)
        rates$chrom = as.character(rates$chrom)
        rates$chrom[is.na(rates$chrom)] = "UNKNOWN"
    }
    
    # convert rates from log-scaled values, so we can multiply by the number of
    # transmissions
    columns = c("syn", "mis", "non", "splice_site", "frameshift")
    rates[, columns] = 10 ** rates[, columns]
    
    # Some genes lack splice site mutation rates (they have NA instead). This
    # impedes estimating the snv.lof.rate, which is estimated from the nonsense
    # and splice_site rates. Convert these NA splice_site values to zero.
    rates$splice_site[is.na(rates$splice_site)] = 0
    
    # get the number of expected mutations, given the number of transmissions
    rates$snv.silent.rate = rates$syn * autosomal
    rates$snv.missense.rate = rates$mis * autosomal
    rates$snv.lof.rate = (rates$non + rates$splice_site) * autosomal
    rates$indel.missense.rate = (rates$frameshift / 9) * autosomal
    rates$indel.lof.rate = rates$frameshift * autosomal
    
    # correct for the known ratio of indels to nonsense, and for transmissions
    # on the X-chromosome
    rates = adjust_indel_rates(rates)
    rates = correct_for_x_chrom(rates, trios$male, trios$female)
    
    # subset to the columns we need to estimate enrichment probabilities
    rates = rates[, c("hgnc", "chrom", "snv.silent.rate", "snv.missense.rate",
        "snv.lof.rate", "indel.missense.rate", "indel.lof.rate")]
    
    return(rates)
}

#' calculate mutation rates based on the gene length
#'
#' This has been discontinued in favor of using rates derived from de novo
#' mutation rates, from Nature Genetics 46:944-950.
#'
#' @param trios list of male and female proband counts in the dataset
#' @export
#'
#' @return a dataframe of mutation rates for genes under different mutation
#'     classes.
get_length_based_rates <- function(trios) {
    
    cds.length = mupit::gene_info$cds_length
    
    # Get the number of autosomal transmissions, so we can estimate the expected
    # number of mutations given the number of potential transmissions.
    autosomal = 2 * (trios$male + trios$female)
    
    # set mutation rates for snvs and indels
    snv_rate = 1.5E-8 # higher than genome-wide mutation rate, due to higher GC ...
    # TODO: check that the following indel mutation rate is correct, since the
    # TODO: comment claims it should be 10% of the SNV mutation rate, but it
    # TODO: currently looks  like 30%
    indel_rate = 0.53E-9 # ~10% of genome-wide SNV mutation rate, no reason to think higher in exome
    
    # specify proportion of coding mutations of different types
    props = list()
    props$snv.lof = 0.0485 # from Daly
    props$snv.missense = 0.6597 # from Daly
    props$snv.silent = 0.2918 # from Daly
    
    props$indel.lof = 0.9 # non-3n, from size distribution in neutral sequence
    props$indel.missense = 0.1
    
    # calculate rates of missense and lof mutations, multiply by 2 for two
    # transmissions per child and number of trios
    rates = data.frame(hgnc = mupit::gene_info$hgnc, chrom = mupit::gene_info$chrom)
    rates$snv.silent.rate = cds.length * snv_rate * props$snv.silent * autosomal
    rates$snv.missense.rate = cds.length * snv_rate * props$snv.missense * autosomal
    rates$snv.lof.rate = cds.length * snv_rate * props$snv.lof * autosomal
    rates$indel.missense.rate = cds.length * indel_rate * props$indel.missense * autosomal
    rates$indel.lof.rate = cds.length * indel_rate * props$indel.lof * autosomal
    
    rates = correct_for_x_chrom(rates, trios$male, trios$female)
    
    return(rates)
}

#' correct mutations rates for sex-chromosome transmission rates
#'
#' @param rates gene-based data frame, containing rates for different mutation
#'     classes.
#' @param male_n number of trios with male offspring
#' @param female_n number of trios with female offspring
#' @export
#'
#' @return a dataframe of mutation rates for genes under different mutation
#'     classes.
correct_for_x_chrom <- function(rates, male_n, female_n) {
    
    # figure out the number of transmissions for autosomal, male and female
    # transmissions
    # TODO: why is male.transmissions equal to the number of females? Why is it
    # TODO: not equal to the number of males?
    autosomal = 2 * (male_n + female_n)
    female = male_n + female_n
    male = female_n
    
    # get scaling factors using the alpha from the most recent SFHS (Scottish
    # Family Health Study) phased de novo data.
    alpha = 3.4
    male_factor = 2 / (1 + (1 / alpha))
    female_factor = 2 / (1 + alpha)
    
    # correct the non-PAR chrX genes for fewer transmissions and lower rate
    # (dependent on alpha)
    chrX = rates$chrom == "X"
    x_factor = ((male * male_factor) + (female * female_factor))/autosomal
    
    rates$snv.silent.rate[chrX] = rates$snv.silent.rate[chrX] * x_factor
    rates$snv.missense.rate[chrX] = rates$snv.missense.rate[chrX] * x_factor
    rates$snv.lof.rate[chrX] = rates$snv.lof.rate[chrX] * x_factor
    rates$indel.missense.rate[chrX] = rates$indel.missense.rate[chrX] * x_factor
    rates$indel.lof.rate[chrX] = rates$indel.lof.rate[chrX] * x_factor
    
    return(rates)
}

#' adapt indel rates for lower rate estimate from validated de novos
#'
#' The indel mutation rates from Samocha et al., Nature Genetics 46:944-950
#' assume that the overall indel mutation rate is 1.25-fold greater than the
#' overall nonsense mutation rate, ie there are 1.25 times as many frameshifts
#' as nonsense mutations. We have our own estimates for the ratio, derived from
#' our de novo validation efforts, which we shall apply in place of the Samocha
#' et al ratios.
#'
#' @param rates data frame of mutation rates.
#' @export
#'
#' @return the rates data frame, with adjusted indel rates.
adjust_indel_rates <- function(rates) {
    
    # I think the following numbers were derived from the DDD dataset
    valid_nonsense = 102
    valid_frameshift = 95
    ddd_ratio = valid_frameshift / valid_nonsense
    samocha_ratio = 1.25  # Nature Genetics 46:944-950 frameshift to nonsense ratio
    
    rates$indel.missense.rate = (rates$indel.missense.rate / samocha_ratio) * ddd_ratio
    rates$indel.lof.rate = (rates$indel.lof.rate / samocha_ratio) * ddd_ratio
    
    return(rates)
}

File: mupit/mutation_rates.py
"""
Copyright (c) 2016 Genome Research Ltd.

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
"""

from __future__ import division

import tempfile
import urllib

import pandas

from mupit.gtf import convert_gtf
from mupit.util import is_url

def get_default_rates(rates_url="http://www.nature.com/ng/journal/v46/n9/extref/ng.3050-S2.xls",
        gencode_url="ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz"):
    """ obtain the table of mutation rates from the Samocha et al paper
    
    Rates for all genes can be obtained from the supplementary material of
    Samocha et al. Nature Genetics 2014 doi:10.1038/ng.3050
    
    Args:
        rates_url: url to supplementary mutation rates table
        gencode_url: url to gencode, or local path. This is required to identify
            chromosomes for the genes in the rates data, since we need to know
            the chromosome in order to corrrect rates on chromosome X.
    
    Returns:
        dataframe of mutation rates, with an extra column for summed lof rate
    """
    
    rates = pandas.read_excel(rates_url, sheetname="mutation_probabilities")
    
    # convert rates from log-scaled values, so we can later multiply by the
    # number of transmissions
    columns = ["syn", "mis", "non", "splice_site", "frameshift"]
    rates[columns] = 10 ** rates[columns]
    
    # sort out the required columns and names.
    rates["hgnc"] = rates["gene"]
    gencode = load_gencode(gencode_url)
    recode = dict(zip(gencode["hgnc"], gencode["chrom"]))
    rates["chrom"] = rates["hgnc"].map(recode)
    
    rates = rates[["hgnc", "chrom", "syn", "mis", "splice_site", "frameshift", "non"]]
    
    return rates

def load_gencode(path):
    """ load gencode table with HGNC symbols and chromosome coordinates
    
    Args:
        path: path to gzipped tab-separated table of gencode gene entries. This
            can be either a url, or local path.
    
    Returns:
        pandas dataframe of HGNC symbols and genome coordiantes
    """
    
    gencode = convert_gtf(path)
    
    # restrict outselves to protein coding genes (or genes which are protein
    # coding in at least some individuals)
    gencode = gencode[gencode["gene_type"].isin(["protein_coding",
        "polymorphic_pseudogene"])]
    
    gencode = gencode[gencode["feature"] == "gene"]
    
    # get the required column names, and strip out all unnecessary columns
    gencode["hgnc"] = gencode["gene_name"]
    gencode["chrom"] = [ x.strip("chr") for x in gencode["seqname"].astype(str) ]
    
    gencode = gencode[["hgnc", "chrom", "start", "end"]].copy()
    
    return gencode

def get_expected_mutations(rates, male, female):
    """ gets numbers of expected mutation per gene
    
    Loads gene-based mutation rates, in order to determine the expected number
    of mutations per gene, given the number of studied probands and adjusts for
    sex-chromosome transmissions.
    
    This defaults to the gene-based mutation rates from Nature Genetics
    46:944-950 (2014) doi:10.1038/ng.3050, but we can pass in other gene-based
    mutation rate datasets.
    
    Args:
        rates: pandas dataframe containing per-gene mutation
        male: number of male probands in the dataset
        female: number of female probands in the dataset
    
    Returns:
        a dataframe of mutation rates for genes under different mutation
        classes.
    """
    
    if rates is None:
        rates = get_default_rates()
    
    autosomal = 2 * (male + female)
    
    expected = rates[["hgnc", "chrom"]].copy()
    
    # get the number of expected mutations, given the number of transmissions
    expected["lof_indel"] = rates["frameshift"] * autosomal
    expected["lof_snv"] = (rates[["non", "splice_site"]].sum(axis=1, skipna=True)) * autosomal
    expected["missense_indel"] = (rates["frameshift"] / 9) * autosomal
    expected["missense_snv"] = rates["mis"] * autosomal
    expected["synonymous_snv"] = rates["syn"] * autosomal
    
    # correct for the known ratio of indels to nonsense, and for transmissions
    # on the X-chromosome
    expected = adjust_indel_rates(expected)
    expected = correct_for_x_chrom(expected, male, female)
    
    # subset to the columns we need to estimate enrichment probabilities
    expected = expected[["hgnc", "chrom", "lof_indel", "lof_snv",
        "missense_indel", "missense_snv", "synonymous_snv"]]
    
    return expected

def correct_for_x_chrom(expected, male_n, female_n):
    """ correct mutations rates for sex-chromosome transmission rates
    
    Args:
        expected: gene-based data frame, containing rates for different mutation
            classes.
         male_n: number of trios with male offspring
         female_n: number of trios with female offspring
    
     Returns:
        a dataframe of mutation rates for genes under different mutation
        classes.
    """
    
    # Calculate the number of transmissions for autosomal, male and female
    # transmissions. The number of transmissions from males is equal to the
    # number of  female probands (since only females receive a chrX from their
    # fathers). Likewise, all offspring receive a chrX from their mothers, so
    # the number of transmissions from females equals the number of probands.
    autosomal = 2 * (male_n + female_n)
    female_transmissions = male_n + female_n
    male_transmissions = female_n
    
    # get scaling factors using the alpha from the most recent SFHS (Scottish
    # Family Health Study) phased de novo data.
    alpha = 3.4
    male_factor = 2 / (1 + (1 / alpha))
    female_factor = 2 / (1 + alpha)
    
    # correct the non-PAR chrX genes for fewer transmissions and lower rate
    # (dependent on alpha)
    chrX = expected["chrom"].isin(["X", "chrX"])
    x_factor = ((male_transmissions * male_factor) + (female_transmissions * female_factor)) / autosomal
    x_factor = pandas.Series([x_factor] * len(chrX), index=expected.index)
    x_factor[~chrX] = 1
    
    expected["missense_snv"] *= x_factor
    expected["missense_indel"] *= x_factor
    expected["lof_snv"] *= x_factor
    expected["lof_indel"] *= x_factor
    expected["synonymous_snv"] *= x_factor
    
    return expected

def adjust_indel_rates(expected):
    """ adapt indel rates for lower rate estimate from validated de novos
    
    The indel mutation rates from Samocha et al., Nature Genetics 46:944-950
    assume that the overall indel mutation rate is 1.25-fold greater than the
    overall nonsense mutation rate, ie there are 1.25 times as many frameshifts
    as nonsense mutations. We have our own estimates for the ratio, derived from
    our de novo validation efforts, which we shall apply in place of the Samocha
    et al ratios.
    
    Args:
        rates: data frame of mutation rates.
    
    Returns:
        the rates data frame, with adjusted indel rates.
    """
    
    # the following numbers were derived from the DDD 4K dataset.
    nonsense_n = 411
    frameshift_n = 610
    ddd_ratio = frameshift_n / nonsense_n
    samocha_ratio = 1.25  # Nature Genetics 46:944-950 frameshift to nonsense ratio
    
    # correct back from using the Samocha et al. ratio
    expected["missense_indel"] /= samocha_ratio
    expected["lof_indel"] /= samocha_ratio
    
    # adjust the indel rates for the DDD indel ratio
    expected["missense_indel"] *= ddd_ratio
    expected["lof_indel"] *= ddd_ratio
    
    return expected
